Generate serialization for NSTextAttachment

sheeparegreat opened the pull request:
#### d6d777b7f57d9ed6d45ad786d13373bfd3494e72
<pre>
Generate serialization for NSTextAttachment
<a href="https://bugs.webkit.org/show_bug.cgi?id=266727">https://bugs.webkit.org/show_bug.cgi?id=266727</a>
<a href="https://rdar.apple.com/119953597">rdar://119953597</a>

Reviewed by Alex Christensen.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::toNSObject):
(WebCore::extractValue):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_addAttachmentForElement):
* Source/WebCore/editing/cocoa/TextAttachmentForSerialization.h: Added.
* Source/WebCore/editing/cocoa/WebCoreTextAttachment.h: Added.
* Source/WebCore/editing/cocoa/WebCoreTextAttachment.mm: Added.
(webCoreTextAttachmentMissingPlatformImage):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/272936@main">https://commits.webkit.org/272936@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3224ea96ba3c79da25090551442ecf71a300302

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33437 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36064 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30381 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14561 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9383 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29505 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29846 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8990 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9112 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29837 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37392 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30368 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35235 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9239 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7199 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33118 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10993 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9824 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4325 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9967 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/editing/cocoa/WebCoreTextAttachment.mm:
```diff
@@ -0,0 +1,39 @@
+
+@implementation NSObject (WebCoreNSTextAttachmentSerialization)
+- (void)setMissingImage:(BOOL)flag {
+    NSNumber *isMissingImage = [NSNumber numberWithBool:flag];
```
This looks like a memory leak.

achristensen07 commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1236,6 +1229,7 @@ static Class _WebMessageDocumentClass()
             ASSERT_WITH_MESSAGE(missingImage != nil, "Unable to find missingImage.");
             attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithData:nil ofType:nil]);
             attachment.get().image = missingImage;
+            [attachment.get() setMissingImage:YES];
```
nit: .get() doesn't add anything here.

achristensen07 commented on Source/WebCore/editing/cocoa/TextAttachmentForSerialization.h:
```diff
@@ -0,0 +1,49 @@
+
+struct TextAttachmentFileWrapper {
+#if !PLATFORM(IOS_FAMILY)
+    bool ignoresOrientation;
```
Let's give this a default initialization to avoid uninitialized memory.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -175,8 +178,28 @@
+        UNUSED_PARAM(value);
+        NSBundle *webCoreBundle = [NSBundle bundleWithIdentifier:@"com.apple.WebCore"];
+#if PLATFORM(IOS_FAMILY)
+        UIImage *missingImage = [PlatformImageClass imageNamed:@"missingImage" inBundle:webCoreBundle compatibleWithTraitCollection:nil];
```
This is duplicate code.  Let's abstract this into a function that returns an image rather than duplicating it.

achristensen07 commented on Source/WebCore/editing/cocoa/WebCoreTextAttachment.h:
```diff
@@ -0,0 +1,46 @@
+@end
+#endif
+
+@interface NSObject (WebCoreNSTextAttachmentSerialization)
```
Adding to NSObject isn't great.  Can't we just add to NSTextAttachment?
Also, I think these at least need a _web_ prefix or something like that to not collide with possible third party applications using the same selector names.
Now that I think about this, adding selectors is probably not great here.  Does checking for pointer equality with the missingImage image work?

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -317,8 +334,23 @@
+            return { { TextAttachmentMissingImage() } };
+        TextAttachmentFileWrapper textAttachment;
+        if (auto accessibilityLabel = [value accessibilityLabel])
+            textAttachment.accessibilityLabel = (String)accessibilityLabel;
```
Casting to a String is strange and probably unnecessary.

achristensen07 commented on Source/WebCore/editing/cocoa/TextAttachmentForSerialization.h:
```diff
@@ -0,0 +1,49 @@
+#if !PLATFORM(IOS_FAMILY)
+    bool ignoresOrientation = false;
+#endif
+    std::optional<String> preferredFilename { };
```
{ } is unnecessary with std::optional, which already has a default initializer to nullopt.
Also, all these types already have a null state, so std::optional looks redundant.

achristensen07 commented on Source/WebCore/editing/cocoa/WebCoreTextAttachment.h:
```diff
@@ -0,0 +1,44 @@
+
+#import <Foundation/Foundation.h>
+
+// Returns an NSImage or UIImage depending on platform
+id webCoreTextAttachmentMissingPlatformImage();
```
Don't we have PlatformImage?

sheeparegreat replied:
Not exactly. I took some inspiration from WKAttachmentTests.mm in the updated PR.

achristensen07 commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -7214,6 +7214,21 @@ enum class WebCore::ContextMenuItemType : uint8_t {
+#if !PLATFORM(IOS_FAMILY)
+    bool ignoresOrientation;
+#endif
+    std::optional<String> preferredFilename;
```
I really don't think these need to be std::optional.  Just check if the String is null.  Same with data, accessibilityLabel.

achristensen07 commented on Source/WebCore/editing/cocoa/WebCoreTextAttachment.h:
```diff
@@ -0,0 +1,50 @@
+
+#import <Foundation/Foundation.h>
+
+#if PLATFORM(MAC)
```
I liked id more than this.

sheeparegreat committed:
Generate serialization for NSTextAttachment
https://bugs.webkit.org/show_bug.cgi?id=266727
rdar://119953597

Reviewed by Alex Christensen.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::toNSObject):
(WebCore::extractValue):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_addAttachmentForElement):
* Source/WebCore/editing/cocoa/TextAttachmentForSerialization.h: Added.
* Source/WebCore/editing/cocoa/WebCoreTextAttachment.h: Added.
* Source/WebCore/editing/cocoa/WebCoreTextAttachment.mm: Added.
(webCoreTextAttachmentMissingPlatformImage):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/272936@main
